includes/class-event.php

```diff
@@ -0,0 +1,399 @@
+	}
+
+	public function get_timestamp(): ?int {
+		return isset( $this->timestamp ) ? $this->timestamp : null;
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	*/
+
+	public function set_status( string $status ): void {
+		$status = strtolower( $status );
+
+		if ( in_array( $status, Events_Store::ALLOWED_STATUSES, true ) ) {
+			$this->status = $status;
+		} else {
+			$this->status = Events_Store::STATUS_PENDING;
+		}
+
+		// Always mark status as changed, as we want to ensure it always sends this data in the SQL if needed.
+		$this->mark_changed( 'status' );
+	}
+
+	public function set_action( string $action ): void {
+		if ( ! empty( $action ) ) {
+			$this->action = $action;
+			$this->mark_changed( 'action' );
+
+			$this->action_hashed = md5( $action );
+			$this->mark_changed( 'action_hashed' );
+		}
+	}
+
+	public function set_args( array $args ): void {
+		$this->args = $args;
+		$this->mark_changed( 'args' );
+
+		$this->instance = self::create_instance_hash( $this->args );
+		$this->mark_changed( 'instance' );
+	}
+
+	public function set_schedule( string $schedule, int $interval ): void {
+		if ( ! empty( $schedule ) && $interval > 0 ) {
+			// Ensure either both or none are set.
+			$this->schedule = $schedule;
+			$this->mark_changed( 'schedule' );
+
+			$this->interval = $interval;
+			$this->mark_changed( 'interval' );
+		}
+	}
+
+	public function set_timestamp( int $timestamp ): void {
+		if ( $timestamp >= 1 ) {
+			$this->timestamp = $timestamp;
+			$this->mark_changed( 'timestamp' );
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Methods for interacting with the object.
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Save the event based on locally-changed props.
+	 *
+	 * @return true|WP_Error true on success, WP_Error on failure.
+	 */
+	public function save() {
+		$changed_data = $this->get_changed_data();
+
+		if ( empty( $changed_data ) ) {
+			// Nothing to save.
+			return new WP_Error( 'cron-control:event:no-save-needed' );
+		}
+
+		if ( ! isset( $this->action, $this->timestamp ) ) {
+			// These are required properties, must be set in order to save.
+			return new WP_Error( 'cron-control:event:missing-props' );
+		}
+
+		$all_row_data = [
+			'status'        => isset( $this->status ) ? $this->status : null,
+			'action'        => $this->action,
+			'action_hashed' => md5( $this->action ),
+			'args'          => serialize( $this->args ),
+			'instance'      => self::create_instance_hash( $this->args ),
+			'timestamp'     => $this->timestamp,
+		];
+
+		if ( isset( $this->schedule ) && isset( $this->interval ) ) {
+			$all_row_data['schedule'] = $this->schedule;
+			$all_row_data['interval'] = $this->interval;
+		}
+
+		// Pick out just the data that has changed.
+		$row_data = [];
```

Comment:
Two main reasons: Seemed overkill to always update every column even if they don't change, and this also lowers the chance of one write contradicting another.

Reply:
I dont think mysql cares too much from a performance stand point if you update only few or more columns.

> lowers the chance of one write contradicting another
Could you explain this more? IF we are worried about parallel execution wouldn't this more likelly result in inconsistent data as some columns might come from one update and some from another?

Reply:
I also like that a call to `$event->save()` will fail if you try to set a prop that was invalid (and thus nothing was marked as updated).

```
$event = new Event( $id );
$event->set_timestamp( -100 );
$result = $event->save();
```

The consumer just needs to check the save result, not all the others methods, to tell if the update was successful.

_____

> If we are worried about parallel execution wouldn't this more likely result in inconsistent data as some columns might come from one update and some from another?

Realistically the only things that will be updated under normal conditions is:

1) The `timestamp` & `schedule/interval` during `->reschedule()` after a job has run.
2) The `status` changes when events are unscheduled.
2) In the future, I want `status` updating more, such as `pending -> running -> pending -> running -> cancelled` for recurring events and `pending -> running -> complete/failed/cancelled` for singles.

The FE requests could be making schedule changes in parallel w/ the runner. My idea is to make it the runner's job to pull from primary DB `srtm` to help account for this. So before it runs it can know for sure if the job is still meant to run, and when it's done running something, before updating back to `pending`, it will check for sure from primary if the job has not been cancelled yet, ensuring it does not overwrite that.

As such, I'd prefer to not update the status in the row unless the consumer calls `set_status()` specifically, w/ the intention of changing the status to help limit the conflict chances w/ the FE.